File: src/my_project/interviews/top_150_questions_round_22/ex_107_maximum_subarray.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod
from collections import deque, defaultdict

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:

        n = len(nums)
        local_maximum = 0
        global_maximum = -(10**10)

        for i in range(n):

            local_maximum = max(nums[i], nums[i]+local_maximum)
            
            if local_maximum > global_maximum:
                global_maximum = local_maximum

        return global_maximum

File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_107_maximum_subarray.ts
function maxSubArray(nums: number[]): number {
    const n = nums.length;
    let localMaximum = 0;
    let globalMaximum = -Infinity;

    for (let i = 0; i < n; i++) {
        localMaximum = Math.max(nums[i], nums[i] + localMaximum);

        if (localMaximum > globalMaximum) {
            globalMaximum = localMaximum;
        }
    }

    return globalMaximum;
};